src/main/java/com/google/googleidentity/oauth2/endpoint/AuthorizationEndpoint.java

```diff
@@ -0,0 +1,91 @@
+/*
+    Copyright 2020 Google LLC
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+package com.google.googleidentity.oauth2.endpoint;
+
+import com.google.googleidentity.oauth2.client.ClientDetailsService;
+import com.google.googleidentity.oauth2.exception.OAuth2ExceptionHandler;
+import com.google.googleidentity.oauth2.exception.OAuth2Exception;
+import com.google.googleidentity.oauth2.util.OAuth2ParameterNames;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import org.apache.http.HttpStatus;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.logging.Logger;
+
+import static com.google.googleidentity.oauth2.exception.OAuth2ExceptionHandler.ErrorCode;
+
+/**
+ * Demo AuthorizationEndpoint for OAuth2 Server
+ */
+@Singleton
+public final class AuthorizationEndpoint extends HttpServlet {
+
+    private static final long serialVersionUID = 5L;
+
+    private static final Logger log = Logger.getLogger("AuthorizationEndpoint");
+
+    private final ClientDetailsService clientDetailsService;
+
+
+    @Inject
+    public AuthorizationEndpoint(ClientDetailsService clientDetailsService) {
+        this.clientDetailsService = clientDetailsService;
+    }
+
+
+    protected void doGet(HttpServletRequest request, HttpServletResponse response)
+            throws ServletException, IOException, UnsupportedOperationException{
+
+        try {
+            String responseType = request.getParameter(OAuth2ParameterNames.RESPONSE_TYPE);
+
+            if (responseType == null) {
+                throw new OAuth2Exception(ErrorCode.NO_RESPONSE_TYPE);
+            }
+            else if (!responseType.equals("token") && !responseType.equals("code")) {
+                throw new OAuth2Exception(ErrorCode.UNSUPPORTED_RESPONSE_TYPE);
+            }
+        }
+        catch(OAuth2Exception exception){
+            log.info(OAuth2ExceptionHandler.getErrorDescription(exception.getErrorCode()));
+            if(OAuth2ExceptionHandler.isRedirectable(exception.getErrorCode())){
```

Comment:
In this case redirect url is required, so we can put this in validator mentioned above as well. This is another advantage using validator, it can make the rest of code simpler by making the promise that all required params are present.

Reply:
put a validator will make a problem for the authorization endpoint.
Here, if we have a valid redirect uri (consistent with client id), we have to do the redirection and send error information to the redirect uri. So actually all other exception should be throw after we have checked the redirect uri(of course client id). If we check all the parameters in a validator first, we may have to throw an exception like no_response_type. At that time, we have not check whether the redirect uri is valid. Therefore we can not decide to redirect or not. The only way is to check the uri in the catch body. There does exist another way, letting the validator do all the checks(including validating redirect uri) and put the check of uri at first. But I am not sure whether this is the kind of validator you talk about(The validator doesn't need to validate the value of each param though, like it doesn't validate if client id and client secret are valid, it just need to ensure these params are set.).
By the way, for other endpoints, I think validators are ok since they do not do redirection, just return json error response.

Reply:
I am fine to check and valid the redirect url first before other checks.